ksql-streams/src/main/java/io/confluent/ksql/execution/streams/StepSchemaResolver.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.streams;
+
+import io.confluent.ksql.execution.plan.AbstractStreamSource;
+import io.confluent.ksql.execution.plan.ExecutionStep;
+import io.confluent.ksql.execution.plan.StreamAggregate;
+import io.confluent.ksql.execution.plan.StreamFilter;
+import io.confluent.ksql.execution.plan.StreamFlatMap;
+import io.confluent.ksql.execution.plan.StreamGroupBy;
+import io.confluent.ksql.execution.plan.StreamGroupByKey;
+import io.confluent.ksql.execution.plan.StreamMapValues;
+import io.confluent.ksql.execution.plan.StreamSelectKey;
+import io.confluent.ksql.execution.plan.StreamSink;
+import io.confluent.ksql.execution.plan.StreamSource;
+import io.confluent.ksql.execution.plan.StreamStreamJoin;
+import io.confluent.ksql.execution.plan.StreamTableJoin;
+import io.confluent.ksql.execution.plan.StreamWindowedAggregate;
+import io.confluent.ksql.execution.plan.TableAggregate;
+import io.confluent.ksql.execution.plan.TableFilter;
+import io.confluent.ksql.execution.plan.TableGroupBy;
+import io.confluent.ksql.execution.plan.TableMapValues;
+import io.confluent.ksql.execution.plan.TableSink;
+import io.confluent.ksql.execution.plan.TableSource;
+import io.confluent.ksql.execution.plan.TableTableJoin;
+import io.confluent.ksql.execution.plan.WindowedStreamSource;
+import io.confluent.ksql.execution.plan.WindowedTableSource;
+import io.confluent.ksql.function.FunctionRegistry;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.util.HandlerMaps;
+import io.confluent.ksql.util.KsqlConfig;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * An interface for computing the schema produced by an execution step, given the schema(s)
+ * going into the step.
+ */
+public final class StepSchemaResolver {
+  private static final HandlerMaps.ClassHandlerMapR2
+      <ExecutionStep, StepSchemaResolver, LogicalSchema, LogicalSchema> HANDLERS
```

Comment:
in the original code, this mapping was tightly coupled with the `SchemaKStream` and now it's defined by this. The approach in this PR means that if we change how any of these steps generate the schema, we need to remember to change this class to (i.e. for structured keys, maybe we want a `StreamGroupBy` to change the key schema of the output - we'd need to update this class too).

Is there any way to make the coupling tighter?

Reply:
@agavra I think this is fair. This class is just glue between the step types and code that computes the schema. We can make it tighter by having the actual schema computation code in the step builders. The supposed advantage of having it looser is that the schema computation is decoupled from the underlying implementation of the steps (e.g. kstreams). Maybe not worth it though. I'll probably change it in a follow up. To be clear, I think it still makes sense to have this class - just that it would work by proxying calls to the step builders.